modules/custom/social_graphql/modules/social_graphql_oauth2/src/Entity/Server.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace Drupal\social_graphql_oauth2\Entity;
+
+use Drupal\Core\TypedData\Exception\MissingDataException;
+use Drupal\graphql\Entity\Server as OriginalServer;
+use Drupal\graphql\GraphQL\Execution\ResolveContext;
+use Drupal\graphql\GraphQL\ResolverRegistryInterface;
+use Drupal\simple_oauth\Authentication\TokenAuthUserInterface;
+use Drupal\simple_oauth\Entity\Oauth2TokenInterface;
+use Drupal\simple_oauth\Oauth2ScopeInterface;
+use GraphQL\Error\UserError;
+use GraphQL\Executor\Values;
+use GraphQL\Type\Definition\ResolveInfo;
+use GraphQL\Type\Definition\WrappingType;
+
+/**
+ * Overrides the GraphQL configuration and request entry point.
+ */
+class Server extends OriginalServer {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getFieldResolver(ResolverRegistryInterface $registry) {
+    $parent = parent::getFieldResolver($registry);
+    if (is_callable($parent)) {
+      $account_proxy = \Drupal::currentUser()->getAccount();
+      $token = NULL;
+      if ($account_proxy instanceof TokenAuthUserInterface) {
+        $token = $account_proxy->getToken();
+      }
+
+      return function ($value, $args, ResolveContext $context, ResolveInfo $info) use ($parent, $token) {
+        $allow_directive_values = $this->getAllowDirectiveValues($info);
+        if ($allow_directive_values) {
+          // If token is not available; it has been removed or a different
+          // authentication is used. In this case we restrict access on
+          // types/fields that have the 'allow' directive.
+          if ($token === NULL) {
+            throw new MissingDataException('There is no OAuth2 token to work on.');
+          }
+          $this->checkAccess($allow_directive_values, $token);
+        }
+        return $parent($value, $args, $context, $info);
+      };
+    }
+    return $parent;
+  }
+
+  /**
+   * Get allow directive values.
+   *
+   * @param \GraphQL\Type\Definition\ResolveInfo $info
+   *   The resolve info object.
+   *
+   * @return array
+   *   Allow directive values array.
+   */
+  private function getAllowDirectiveValues(ResolveInfo $info): array {
+    $allow_directives = [
+      'allowUser',
+      'allowBot',
+      'allowAll',
+    ];
+    $directive_values = [];
+
+    foreach ($allow_directives as $allow_directive) {
+      $directive_def = $info->schema->getDirective($allow_directive);
+
+      if ($directive_def !== NULL) {
+        $type = $info->returnType;
+        if ($type instanceof WrappingType) {
+          $type = $type->getWrappedType(TRUE);
+        }
+
+        // Get directive values on the type.
+        $type_directive_values = $type->astNode !== NULL ? Values::getDirectiveValues(
+          $directive_def,
+          $type->astNode
+        ) : NULL;
+
+        // Get directive values on the field.
+        $field_directive_values = $info->fieldDefinition->astNode ? Values::getDirectiveValues(
```

Comment:
Prefer explicit checks on non-boolean values over using PHP's truthy behavior (I thought this was a lint rule somwhere 🤔 )

```suggestion
        $field_directive_values = $info->fieldDefinition->astNode !== NULL ? Values::getDirectiveValues(
```